### DISCUSSION TITLE
Low FastAPI performance in test

### DISCUSSION THREAD

**Discussion by eddy2by4** [fastapi/fastapi#13115] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Source code can be found :

https://github.com/antonputra/tutorials/tree/232/lessons/232/python-app
```

### Description

There is this video comparing FastAPI to Go.

https://www.youtube.com/watch?v=sxdpKG-6HSY

And I don't believe the DB latencies in that test are normal:
80+ ms latency on db save
40+ ms latency on redis set 

Is FastAPI that slow or is there something wrong with the setup ?

### Operating System

Linux

### FastAPI Version

0.115.6

### Pydantic Version

2.10.3

### Python Version

3.13

**Comment by eddy2by4** [comment]
Some thoughts here:

Python ( https://www.youtube.com/watch?v=sxdpKG-6HSY ):
Without any DB interaction: 11k req/s
With Postgresql + Redis: 2,5k req/s 

Nodejs - same setup ( https://www.youtube.com/watch?v=UbDRSZ3IK2Y )
Without any DB interaction: 45k req/s
With Postgresql + Redis: 12k req/s

Why is the overhead for Python using Postgresql + Redis so big? 
Its clearly not the server since he can do 11k without DB, but its clearly not the DB/Redis engines, since other languages can use them to have up to 12k req/s 

I would expect personally, that giving the same db/cache servers, the languages to perform similarly, but its clearly a big difference.
Is asyncpg, aioredis so much slower than nodejs's libraries for DB/cache interactions?

**Comment by Kludex** [comment]
Maybe the dependency injection, or the non async dependencies, or the orjson. I didn't check the benchmark, just randomly some files there.

I recommend checking my FastAPI tips repository. I'm on my phone, so I can't provide the link easily.

**Reply by eddy2by4** [reply]
Hmm,
I don't see a problem with the dependency injection, but I am not so proficient as you in the matter.
The DB functions are indeed async, but the get_devices() is not async, do you think combining async and not async functions might affect performance overall ?

What about orjson?
When you have a few minutes, can you please take a look on the source code for some more insight?

Don't you also believe, that the performance of Fastapi with the db/redis should have been closer to the nodejs one?

Thanks !

**Reply by eddy2by4** [reply]
@Kludex I've opened an issue with some things I've found for the maintainer:

https://github.com/antonputra/tutorials/issues/397

Please add there your findings also, so we can track.
